### PR TITLE
jefftmarks/APPEALS-22287

### PR DESCRIPTION
**Pull request by jefftmarks**
Resolves [APPEALS-22287](https://vajira.max.gov/browse/APPEALS-22287)

# Description
There are a number of conditional column constraints within the schema outlined for our Package Manger effort ([APPEALS-51](https://vajira.max.gov/browse/APPEALS-51)). In order to enforce these, there should be [ActiveRecord validators](https://guides.rubyonrails.org/v5.2/active_record_validations.html) implemented into our new models.

The constraints can be viewed in [our 4+1](https://app.mural.co/t/bahdigitalexperience6902/m/bahdigitalexperience6902/1675100430669/ad5530ce43f88cd86bfd6e2793f2bc539c4494f1?sender=acd2468a-7e7b-4a47-a604-8cc2519cae84) in the Process and Physical views.

## Acceptance Criteria
- [x] Code compiles correctly
1. VbmsCommunicationPackage
    - [x] file Number is required
    - [x] name length must not exceed 255 characters
    - [x] "user friendly" name matches regex: ^[\w !*+,-.:;=?] {0,255}$.
    - [x] copies must not be null, and be between 1-500 (inclusive)
2. VbmsDistribution
    - [x] communication_package_id: Must not be null, and needs to match with a primary key of a VbmsCommunicationPackage
    - [x] recipient: must not be null
    - [x] recipient_type: Must be one of 'organization', 'person', 'system' or 'ro-colocated'
    - [x] name: If recipient_type == System, ro-colocated or Organization then must not be null
    - [x] first_name: If recipient_type == person then must not be null
    - [x] last_name: If recipient_type == person then must not be null
    - [x] poa_code: if recipient_type == ro-colocated then Must not be null
    - [x] claimant_station_of_jurisdiction: if recipient_type == ro-colocated then Must not be null
3. VbmsDistributionDestination
    - [x] destination_type: Must be 'domesticAddress', 'internationalAddress', 'militaryAddress', 'derived', 'email'. Cannot be 'physicalAddress'.
    - [x] address_line_1: If destination_type == domestic, international, or military then must not be null
    - [x] address_line_2: If treat_line_2_as_addressee == true then must not be null
    - [x] address_line_3: if treat_line_3_as_addressee == true then Must not be null
    - [x] treat_line_2_as_addressee cannot be false if treat_line_3_as_addressee  == true
    - [x] city: If destination_type == domestic, international, or military then must not be null
    - [x] state: Must be exactly two-letter ISO 3166-2 code.
    - [x] If destination_type == domestic or military then must not be null
    - [x] postal_code: If destination_type == domestic or military then must not be null
    - [x] country_name: if destination_type == international then must not be null
    - [x] country_code: Must be exactly two-letter ISO 3166 code if provided.
    - [x] email_address: If destination_type == email then must not be null
    - [x] phone_number: If destination_type == sms then must not be null
    - [x] vbms_distribution_id: Must not be null
    - [x] Should correspond with a primary key of a VbmsDistributionDestination

## Testing Plan
Go to [APPEALS-22807](https://vajira.max.gov/browse/APPEALS-22807)

### Test Coverage
Did you include any test coverage for your code? Check below:
- [x] RSpec
- [ ] Jest
- [ ] Other

### CODE REVIEW

File: spec/models/vbms_distribution_destination_spec.rb [hunk 1]
```diff
@@ -1,5 +1,136 @@
...
+      expect(destination).to_not be_valid
+    end
+
+    it "is not valid without a city" do
+      destination.city = nil
+      expect(destination).to_not be_valid
+    end
+
+    it "is not valid without a country code" do
+      destination.country_code = nil
```
Comment (ThorntonMatthew) [inline spec/models/vbms_distribution_destination_spec.rb:65]: Do you see any value in checking for specific validation failure messages within `<model_obj>.errors.messages` to ensure particular validations are being targeted by the tests? I am leaning towards that being overkill.

**Review by mchbidwell — COMMENTED** [review]
change/update 'document_reference' to 'copies'
add check for when line 2AsAddress is null while line 3AsAddress is true (contain address)

File: spec/models/vbms_distribution_destination_spec.rb (hunk 1 above)
Comment (jefftmarks) [reply spec/models/vbms_distribution_destination_spec.rb:65]: Is something like this what you have in mind?
```
it "is not valid without a filenumber" do
    package.file_number = nil
    expect(package).to_not be_valid
    expect(package.errors[:file_number]).to eq(["can't be blank"])
end
```

Comment (jefftmarks) [reply spec/models/vbms_distribution_destination_spec.rb:65]: Definitely not overkill... because I just caught an error!

**Review by ThorntonMatthew — APPROVED** [review]
🚀 Nice work!

**Pull request merged**